refactor: replace RecursiveMutex `m_most_recent_block_mutex` with Mutex

theStack opened the pull request:
This PR is related to #19303 and gets rid of the RecursiveMutex `m_most_recent_block_mutex`. All of the critical sections (5 in total) only directly access the guarded elements, i.e. it is not possible that within one section another one is called, and we can use a regular Mutex:

https://github.com/bitcoin/bitcoin/blob/b019cdc036343a437fd7ced85467bd95f48d84c4/src/net_processing.cpp#L1650-L1655

https://github.com/bitcoin/bitcoin/blob/b019cdc036343a437fd7ced85467bd95f48d84c4/src/net_processing.cpp#L1861-L1865

https://github.com/bitcoin/bitcoin/blob/b019cdc036343a437fd7ced85467bd95f48d84c4/src/net_processing.cpp#L3149-L3152

https://github.com/bitcoin/bitcoin/blob/b019cdc036343a437fd7ced85467bd95f48d84c4/src/net_processing.cpp#L3201-L3206

https://github.com/bitcoin/bitcoin/blob/b019cdc036343a437fd7ced85467bd95f48d84c4/src/net_processing.cpp#L4763-L4769

The scope of the last critical section is reduced in the first commit, in order to avoid calling the non-trivial method `CConnman::PushMessage` while the lock is held.

hebasto commented:
Concept ACK. Thanks!

hebasto commented on src/net_processing.cpp:
```diff
@@ -4704,7 +4704,7 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
                     bool fGotBlockFromCache = false;
                     {
-                        LOCK(cs_most_recent_block);
+                        LOCK(most_recent_block_mutex);
```
> ... it is not possible that within one section another one is called...

It is not obvious where functions with a non-trivial body--like `PushMessage`--are called. Even everything is correct now, there is a possibility to introduce a recursion in he future by modifying the `PushMessage` function or other functions been called from within it.

Could we just call `PushMessage` without holding `most_recent_block_mutex`?

theStack replied:
Fair enough, will take a deeper look.

hebasto reviewed (commented):
Approach ACK cac83660c07568820dcd0cd9cc71548bf05857c3.

theStack committed:
refactor: reduce scope of lock `m_most_recent_block_mutex`

This avoids calling the non-trivial method
`CConnman::PushMessage` within the critical section.

theStack commented:
Added another commit which reduces the scope of the lock `most_recent_block_mutex`, [as suggested by @hebasto](https://github.com/bitcoin/bitcoin/pull/24062#discussion_r784643490). Rather than calling the non-trivial method `CConnman::PushMessage` immediately, only the (cached) cmpctblock message is now assembled and sent after, outside of the critical section.

theStack committed:
refactor: replace RecursiveMutex `m_most_recent_block_mutex` with Mutex

In each of the critical sections, only the the guarded variables are
accessed, without any chance that within one section another one is
called.  Hence, we can use an ordinary Mutex instead of RecursiveMutex.

theStack commented:
Rebased on master. Note that the scripted-diff renaming commit (getting rid of the lock's `cs_` prefix) was dropped as this was already resolved in the recently merged PR #24543. Also updated the PR title and description accordingly with the current critical section code snippets.

theStack commented:
Rebased on master, updated the PR description with the current critical section code snippets (shorter after #20799 has been merged). The boolean variable `fGotBlockFromCache` is now removed in the first commit and a std::optional on the cached message is used instead, which I think is a bit more elegant.

furszy reviewed (approved):
Code ACK 83003ffe with a small comment.

hebasto reviewed (approved):
ACK 83003ffe049a432f6fa4127e054f073127e70b90

w0xlt reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/24062/commits/83003ffe049a432f6fa4127e054f073127e70b90

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -4759,15 +4759,16 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
                     }
-                    if (!fGotBlockFromCache) {
+                    if (cached_cmpctblock_msg.has_value()) {
+                        m_connman.PushMessage(pto, std::move(cached_cmpctblock_msg.value()));
```
Just realized, all of this could be written as:

```cpp
if (WITH_LOCK(m_most_recent_block_mutex, return m_most_recent_block_hash == pBestIndex->GetBlockHash())) {
    m_connman.PushMessage(pto, msgMaker.Make(NetMsgType::CMPCTBLOCK, *m_most_recent_compact_block));

hebasto replied:
`m_most_recent_compact_block` should also be placed within a critical section.
